Use DIExpression::foldConstantMath at the result of a Salvaged expression

rastogishubham opened the pull request:
This patch uses `DIExpression::foldConstantMath()` at the result of a Salvaged expression, that is, it runs the folding optimizations after an expression has been salvaged completely, to reduce how many times the fold optimization function is called. Which should help in reducing the size of DIExpressions that grow because of salvaging debug info

After checking the size of the dSYM with and without this change, I saw a decrease of about 300KB, where the debug_loc section is about 1.6 GB in size.

Where the debug loc section reduced in size by 212KB and it is 193MB in size, the rest comes from the debug_info section

This is part of a stack of patches and comes after:
https://github.com/llvm/llvm-project/pull/69768
https://github.com/llvm/llvm-project/pull/71717
https://github.com/llvm/llvm-project/pull/71718
https://github.com/llvm/llvm-project/pull/71719

jmorse reviewed (approved):
LGTM with performance consideration mentioned on the parent patch

rastogishubham committed:
Use DIExpression::foldConstantMath at the result of a Salvaged expression

The functions salvageDbgAssignAddress, and salvageDebugInfoForDbgValues
call into DIExpression::AppendOpsToArg, many times, use
DIExpression::foldConstantMath at the end of the salvaging loop to
fold any constant math to reduce the size of the salvaged expression.

rastogishubham commented:
@jmorse @adrian-prantl patch has been updated, I am unsure about what to do for DebugInfo/salvage-limit-expr-size.ll, because of this patch, there is no longer a limit of 128 salvageable instructions, because they get folded down to one instruction

adrian-prantl commented on llvm/lib/IR/DebugInfoMetadata.cpp:
```diff
@@ -1774,7 +1774,8 @@ DIExpression *DIExpression::appendOpsToArg(const DIExpression *Expr,
            "Location Index must be 0 for a non-variadic expression.");
     SmallVector<uint64_t, 8> NewOps(Ops.begin(), Ops.end());
-    return DIExpression::prependOpcodes(Expr, NewOps, StackValue);
+    return DIExpression::prependOpcodes(Expr, NewOps, StackValue)
```
Can prependOpcodes return nullptr?

SLTozer commented:
> I am unsure about what to do for DebugInfo/salvage-limit-expr-size.ll, because of this patch, there is no longer a limit of 128 salvageable instructions, because they get folded down to one instruction

In some ways, this is quite a good problem to have! I'm not well-acquainted with this patch, but if there's any kind of legal DIExpression that won't get constant-folded then you could substitute that; it looks like you've not covered all the logical operators (not, xor, etc), there are some operators that might be effectively repeatable no-ops (DW_OP_rot), and if all else fails then it should be possible to repeat `DW_OP_LLVM_arg, 0` in the expression, so that you're effectively performing `x + x + x + x + x + x +...` until you have a DIExpression of the right length.

SLTozer commented:
Looking at some compile time stats as-of the most recent rebase: this patch stack has a geomean +0.32% to [compile time](http://llvm-compile-time-tracker.com/compare.php?from=1f01c580444ea2daef67f95ffc5fde2de5a37cec&to=b4d7e1cb2a81c465afe3d1cdae5712b393cc1ce9&stat=instructions%3Au) at ReleaseLTO-g, in return for -0.01% [file size](http://llvm-compile-time-tracker.com/compare.php?from=1f01c580444ea2daef67f95ffc5fde2de5a37cec&to=b4d7e1cb2a81c465afe3d1cdae5712b393cc1ce9&stat=size-file). Looking at the specific projects, it looks as though the best cost-benefit ratio comes from tramp3d-v4 (which probably generates a higher proportion of constant-foldable expressions), which has +0.53% compile time for -0.05% file size; I think that's still not a _great_ ratio. I'd be interested to see if there are any heuristics that could be applied to reduce the compile time increase.

A few suggestions that might improve performance:
- Move this from appendOpsToArg to salvageDebugInfoImpl - that's where I'd guess all of the foldable expressions are coming from, and in addition to being called less often, there's contextual information there that could help simplify further, such as...
- Only check for a constant foldable expression if we have just salvaged a foldable operation with a constant operand - any other salvage operation can't have turned a non-foldable expression to a foldable one.
- Instead of performing this on every DIExpression that gets updated, which in many cases will actually be performing this on the same DIExpression repeatedly and returning the same result each time, you could manually track the users of each updated DIExpression, then for each DIExpression try to constant-fold it _once_ and update each user to point to the new DIExpression.

I think that simplifying DIExpressions _is_ something that should be pursued, so I apologize for adding to the workload of actually implementing it!

rastogishubham commented:
@SLTozer Thanks for the feedback! I really appreciate it, I will try to update the patches with the suggestions

rastogishubham commented:
@SLTozer After taking your feedback into account, I believe I have a solution that is more appropriate. [The compile time](https://llvm-compile-time-tracker.com/compare.php?from=82a8c1cf35e6794d4d1e56797d58abbed0112ad9&to=abbe25add5dc055405f84b59f8baa7d8b2faf795&stat=instructions%3Au) is not significanlty higher, and the [file size](https://llvm-compile-time-tracker.com/compare.php?from=82a8c1cf35e6794d4d1e56797d58abbed0112ad9&to=abbe25add5dc055405f84b59f8baa7d8b2faf795&stat=size-file) reduction is the same.

I would like to note that the size of the file will go down, because currently `DIExpression::foldConstantMath` doesn't fold signed operations, and doesn't fold the logical operations either. Please let me know what you think

adrian-prantl commented on llvm/lib/IR/DebugInfoMetadata.cpp:
```diff
@@ -2022,6 +2025,349 @@ DIExpression::constantFold(const ConstantInt *CI) {
+  assert(Result->isValid() && "concatenated expression is not valid");
+  return Result;
+}
+
```
It might make sense to move these functions into a new DIExpressionOptimizer.cpp file if plan to add more rules. WDYT?

rastogishubham replied:
I think that makes sense!

adrian-prantl commented on llvm/lib/IR/DebugInfoMetadata.cpp:
```diff
@@ -2022,6 +2025,349 @@ DIExpression::constantFold(const ConstantInt *CI) {
+    auto Op6Raw = Op6->getOp();
+    if (tryFoldCommutativeMathWithArgInBetween(Op1Raw, Op1Arg, Op2Raw, Op3Raw,
+                                               Op4Raw, Op5Raw, Op5Arg, Op6Raw,
+                                               Loc, Cursor, ResultOps))
```
Wouldn't it make the code shorter/readable if we instead of all getOp() calls were changing the rule implementations to have a single ArrarRef<UInt64> argument that is guaranteed to have size N?

adrian-prantl replied:
In a previous conversation you mentioned that this wouldn't be possible because not every Operation has an Argument. And while it looks like the implementation of getArg doesn't care about this detail,
```
    uint64_t getArg(unsigned I) const { return Op[I + 1]; }

    unsigned getNumArgs() const { return getSize() - 1; }
```
it still is necessary to call getArg etc, to know where the next operation starts.

So would it make sense to convert the DIExpression suffix/prefix we are operating on into an array of ExprOperand at the beginning, and hide that in a separate helper function, so this function becomes much shorter?

rastogishubham replied:
Yes I agree with that. I think an array of `ExprOperand` makes sense

MaskRay commented on llvm/include/llvm/IR/DIExpressionOptimizer.h:
```diff
@@ -0,0 +1,81 @@
+//===- llvm/IR/DIExpressionOptimizer.h - Constant folding of DIExpressions --*-
```
80 columns and avoid wrapping

You can remove `llvm/IR/` to get more columns.

rastogishubham replied:
Umm pretty sure the `+1,81` refers to the number of lines, not columns

MaskRay commented on llvm/include/llvm/IR/DIExpressionOptimizer.h:
```diff
@@ -0,0 +1,81 @@
+
+#include "llvm/IR/DebugInfoMetadata.h"
+
+using namespace llvm;
```
We use `using namespace llvm;` in .cpp files but avoid them in `.h`.

Instead, wrap the declarations in `namespace llvm {`. See neighbor files for convention.

MaskRay commented on llvm/lib/IR/DIExpressionOptimizer.cpp:
```diff
@@ -0,0 +1,281 @@
+bool operationsAreFoldableAndCommutative(uint64_t Op1, uint64_t Op2) {
+  if (Op1 != Op2)
+    return false;
+  switch (Op1) {
```
`return xx == yy || xx == zz;`

MaskRay commented on llvm/include/llvm/IR/DIExpressionOptimizer.h:
```diff
@@ -0,0 +1,81 @@
+//
+//===----------------------------------------------------------------------===//
+//
+// Declarations for functions to constant fold DIExpressions.
```
This can use more descriptions.

 The similar comment in the .cpp file might not be that useful.

rastogishubham commented:
@SLTozer Just pinging again in case you missed the updates,
I think I have addressed the compile time issues here: https://github.com/llvm/llvm-project/pull/71721#issuecomment-2099006338

SLTozer commented:
> @SLTozer Just pinging again in case you missed the updates,

Hi - just letting you know that I've been preoccupied by other work, but I'm looking at this now; the improvements in compile time LGTM. Outside of that my specific approval to land isn't necessary as long as the patches receive general approval, but otherwise I'll work my way through the patches that haven't yet got the green tick.

adrian-prantl commented:
To summarize an offline conversation I had with @rastogishubham — I think hardcoding the rules like this is a great starting point.

Going forward (i.e., _after_ this patch) I would like to see us at least explore if we can write the rules in a declarative form where they are compiled into function (similar to the rule bodies right now) and some sort of type signature that conceptually consists of the parameters each rules consumes (e.g.: `uconst, uconst, operator`) and a rewrite system that abstracts the the input DIExpression into its type signature and then matches the rules accordingly, so we don't need to hardcode the application in foldConstantMath(). So for example, the rules themselves could be stored in a tree where the edges are types and the rules are stored in the nodes:
```
          []
  const  / \ op
        []  ...
 const / 
      []
       \ op
       []  all rules that consume const,const,op
 ```
Maybe that tree can be stored in linear fashion and even produced at compile time by tablegen? Maybe this is also over-engineering the problem ;-)

jryans commented:
I agree with @adrian-prantl, a rewrite rule design should work quite well here. It should make this system easier to maintain and understand for long term.

Of course, that’s future work for after this PR, but I am happy to hear people are thinking along these lines. 😄

SLTozer reviewed (commented):
> Maybe that tree can be stored in linear fashion and even produced at compile time by tablegen? Maybe this is also over-engineering the problem ;-)

Maybe just the tiniest bit 😄! But I do agree that we should have a more extensible approach to this, which could be tablegen, or any other simple declarative style.
With regards to the current design, it seems suitable for a first pass. In fact, if I were to suggest a further (non-arithmetic) pattern, I think that replacing patterns of consecutive `DW_OP_LLVM_convert` would be a very effective way to reduce expression sizes during compilation - off the top of my head I don't believe the "types" mean anything, we just emit them as truncations (e.g. `DW_OP_constu 0x0000ffff, DW_OP_and`) at the end of compilation, and we sometimes end up with a lot of converts after salvaging, which results in slower expression processing and some debug values being dropped. That pattern doesn't need to/probably shouldn't appear in this patch, though!

rastogishubham commented on llvm/lib/IR/DIExpressionOptimizer.cpp:
```diff
@@ -0,0 +1,349 @@
+    if (!Op)
+      break;
+    auto OpRaw = Op->getOp();
+    auto OpArg = Op->getArg(0);
```
That is a great catch! Thank you so much!

rastogishubham commented:
@SLTozer Thanks for the review

@adrian-prantl thanks for summarizing the discussion we had!
